File: qpi-driver/js/src/builtins/bluefors-gen1.ts
/**
 * Cryostat monitor for Bluefors Control Software Gen. 1 (RFC 0001 §7), the
 * TypeScript counterpart of the Python `qpi_driver.builtins.bluefors_gen1`
 * driver.
 *
 * Note the Gen. 1: Bluefors also ships a Gen. 2 Control Software with its own
 * API, which is out of scope here and would need its own driver.
 *
 * A report-only driver: it never handles `JobDispatch` (it is not a QPU), it
 * polls the Bluefors Remote Access Control API Gen. 1 "values" endpoint on a
 * timer and emits a `CryostatReading` event with whatever it read. The value
 * tree path for a channel (e.g. `mapper.bf.tmc`) is the endpoint path with dots
 * replaced by slashes: `GET {baseUrl}/values/mapper/bf/tmc`.
 *
 * It lives in its own sub-module so it is only bundled when imported —
 * `import { BlueforsGen1Driver } from "qpi-driver/builtins/bluefors-gen1"` — the
 * bundler equivalent of the Python `qpi-driver[bluefors_gen1]` extra.
 */

import { Event, EventType } from "../events.js";
import { QpiDriver, type QpiDriverOptions } from "../driver.js";

export const DEFAULT_BASE_URL = "http://127.0.0.1:49099";
export const DEFAULT_POLL_INTERVAL_MS = 5000;
export const DEFAULT_TIMEOUT_MS = 5000;

/** Options for the Bluefors Gen. 1 monitor, on top of the base driver options. */
export interface BlueforsGen1Options extends QpiDriverOptions {
  /** Base URL of the Bluefors Control API. Defaults to {@link DEFAULT_BASE_URL}. */
  blueforsBaseUrl?: string;
  /**
   * Channels to poll: either a `{ path: unit }` map (e.g.
   * `{ "mapper.bf.tmc": "K" }`) or a bare list of paths. The Bluefors basic
   * read response does not report units, so they are supplied here.
   */
  channels?: Record<string, string> | string[];
  /** Optional Bluefors access key, sent as the `key` query parameter. */
  apiKey?: string;
  /** Milliseconds between polls. Defaults to {@link DEFAULT_POLL_INTERVAL_MS}. */
  pollIntervalMs?: number;
  /** Per-channel HTTP timeout in ms. Defaults to {@link DEFAULT_TIMEOUT_MS}. */
  timeoutMs?: number;
}

/** One channel's latest value in the emitted payload. */
export interface Reading {
  value: number | null;
  unit: string;
  status: string;
}

/** Polls Bluefors Gen. 1 Control API channels and emits readings on a timer. */
export class BlueforsGen1Driver extends QpiDriver {
  private readonly baseUrl: string;
  private readonly channels: Record<string, string>;
  private readonly apiKey: string;
  private readonly timeoutMs: number;

  constructor(options: BlueforsGen1Options) {
    super(options);
    this.baseUrl = (options.blueforsBaseUrl ?? DEFAULT_BASE_URL).replace(
      /\/+$/,
      "",
    );
    this.channels = normalizeChannels(options.channels);
    this.apiKey = options.apiKey ?? "";
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;

    this.every(options.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS, () =>
      this.poll(),
    );
  }

  /**
   * Ignore every inbound event: the monitor only reports upward. It never
   * handles `JobDispatch`; it is a separate driver from the QPU (RFC 0001 §4).
   */
  handleEvent(event: Event): void {
    console.warn(
      `[bluefors] dropping event ${event.id}: monitor does not handle ${event.type}`,
    );
  }

  /**
   * Read every configured channel and emit whatever succeeded. A channel that
   * fails to read is recorded with a null value and an ERROR status rather than
   * aborting the tick; if every channel fails, nothing is emitted.
   */
  async poll(): Promise<void> {
    const readings: Record<string, Reading> = {};
    for (const [channel, unit] of Object.entries(this.channels)) {
      readings[channel] = await this.readChannel(channel, unit);
    }

    const anyOk = Object.values(readings).some((r) => r.status !== "ERROR");
    if (!anyOk) {
      console.warn(
        `[bluefors] all ${Object.keys(readings).length} channel(s) failed this tick; skipping emit`,
      );
      return;
    }

    this.emit(new Event(EventType.CryostatReading, { readings }));
  }

  private async readChannel(channel: string, unit: string): Promise<Reading> {
    const path = channel.replaceAll(".", "/");
    const url = new URL(`${this.baseUrl}/values/${path}`);
    if (this.apiKey) {
      url.searchParams.set("key", this.apiKey);
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(url, { signal: controller.signal });
      if (!resp.ok) {
        throw new Error(`status ${resp.status}`);
      }
      const body = (await resp.json()) as BlueforsResponse;
      const content = body?.data?.content ?? {};
      const sample = content.latest_valid_value ?? content.latest_value ?? {};
      return {
        value: parseValue(sample.value),
        unit,
        status: sample.status ?? "UNKNOWN",
      };
    } catch (err) {
      console.error(`[bluefors] failed to read channel ${channel}:`, err);
      return { value: null, unit, status: "ERROR" };
    } finally {
      clearTimeout(timer);
    }
  }
}

interface BlueforsSample {
  value?: unknown;
  status?: string;
}

interface BlueforsResponse {
  data?: {
    content?: {
      latest_valid_value?: BlueforsSample;
      latest_value?: BlueforsSample;
    };
  };
}

/**
 * Coerce a raw Bluefors value (which may be a number or a numeric string) to a
 * number, or null when it is absent or non-numeric.
 */
function parseValue(raw: unknown): number | null {
  if (raw === null || raw === undefined || raw === "") {
    return null;
  }
  const value = typeof raw === "number" ? raw : Number(raw);
  return Number.isFinite(value) ? value : null;
}

/** Turn a channel list/map/undefined into the `{ path: unit }` shape used internally. */
export function normalizeChannels(
  channels: Record<string, string> | string[] | undefined,
): Record<string, string> {
  if (!channels) {
    return {};
  }
  if (Array.isArray(channels)) {
    return Object.fromEntries(channels.map((path) => [path, ""]));
  }
  return { ...channels };
}

/** Parse `"path[:unit],path[:unit],..."` into a channel->unit map. */
export function parseChannels(raw: string): Record<string, string> {
  const channels: Record<string, string> = {};
  for (const part of raw.split(",")) {
    const trimmed = part.trim();
    if (!trimmed) {
      continue;
    }
    const [path, unit = ""] = trimmed.split(":");
    channels[path.trim()] = unit.trim();
  }
  return channels;
}

File: qpi-driver/py/qpi_driver/builtins/bluefors_gen1.py
"""Cryostat monitor for Bluefors Control Software Gen. 1 (RFC 0001 §7, Phase 3).

Note the *Gen. 1*: Bluefors also ships a Gen. 2 Control Software with its own
API, which is out of scope here and would need its own driver if supported.

A driver that only reports upward: it never handles ``JobDispatch`` (it is
not a QPU), it just polls the Bluefors Remote Access Control API Gen. 1's HTTP
``values`` endpoint on a timer and emits a ``CryostatReading`` event with
whatever it read (Bluefors Remote Access Control API Gen. 1 Technical
Reference, "Program structure" and Appendix I: ``values``).

The value tree path for a channel (e.g. ``mapper.bf.tmc``) is exactly what the
Bluefors API endpoint path is, with dots replaced by slashes, e.g.
``GET {base_url}/values/mapper/bf/tmc``. Which channels exist, and their
names, depend on how a given system's mappers are configured, so they are
supplied as configuration rather than hard-coded.

Ships as an officially maintained driver — install with
``qpi-driver[cli,bluefors_gen1]`` and run with ``qpi-driver monitor --device
bluefors_gen1`` (see ``qpi_driver.cli``), the same tier as the qblox/quantify
executors.
"""

import logging
from typing import Any

import requests

from qpi_driver.builtins.qpu import _normalize_qpi_addr
from qpi_driver.events import Event, EventType
from qpi_driver.sdk import DEFAULT_RECV_TIMEOUT_MS, QpiDriver

log = logging.getLogger(__name__)

DEFAULT_POLL_INTERVAL = 5.0
DEFAULT_TIMEOUT = 5.0


class BlueforsGen1Driver(QpiDriver):
    """Polls Bluefors Gen. 1 Control API channels and emits readings on a timer.

    Attributes:
        bluefors_base_url: Base URL of the Bluefors Control API, e.g.
            ``http://localhost:49099``.
        channels: Maps a value-tree channel path (e.g. ``"mapper.bf.tmc"``) to
            a display unit (e.g. ``"K"``), which the Bluefors API's basic read
            response does not itself report. An empty unit is fine.
        api_key: Optional Bluefors API access key, sent as the ``key`` query
            parameter (Bluefors reference §3.5.1).
        poll_interval: Seconds between polls.
        timeout: HTTP timeout per channel read, in seconds.
    """

    OPERATION = "monitor"

    def __init__(
        self,
        qpi_addr: str = "http://127.0.0.1:8090",
        token: str = "",
        name: str = "bluefors-gen1-monitor",
        bluefors_base_url: str = "http://127.0.0.1:49099",
        channels: dict[str, str] | list[str] | None = None,
        api_key: str = "",
        poll_interval: float = DEFAULT_POLL_INTERVAL,
        timeout: float = DEFAULT_TIMEOUT,
        ca_fingerprint: str = "",
        ca_file_path: str = "./bin/qpi.ca.pem",
        recv_timeout_ms: int = DEFAULT_RECV_TIMEOUT_MS,
    ) -> None:
        super().__init__(
            qpi_addr=_normalize_qpi_addr(qpi_addr),
            token=token,
            name=name,
            ca_fingerprint=ca_fingerprint,
            ca_file_path=ca_file_path,
            recv_timeout_ms=recv_timeout_ms,
        )
        self.bluefors_base_url = bluefors_base_url.rstrip("/")
        self.channels = normalize_channels(channels)
        self.api_key = api_key
        self.poll_interval = poll_interval
        self.timeout = timeout

        self.every(self.poll_interval, self._poll)

    def handle_event(self, event: Event) -> None:
        """Ignore every inbound event — the monitor only reports upward.

        It never handles ``JobDispatch``; it is a separate driver from the
        QPU, not part of it (RFC 0001 §4).
        """
        log.warning(
            "dropping event %s: bluefors_gen1 driver does not handle %s",
            event.id,
            event.type.value,
        )

    def _poll(self) -> None:
        """Read every configured channel and emit whatever succeeded.

        A channel that fails to read (timeout, HTTP error, unexpected shape)
        is recorded with a ``None`` value and an ``ERROR`` status rather than
        raising, so one bad channel does not lose the rest of the tick. If
        every channel fails, nothing is emitted for this tick.
        """
        readings: dict[str, dict[str, Any]] = {}
        for channel, unit in self.channels.items():
            readings[channel] = self._read_channel(channel, unit)

        if not any(r["status"] != "ERROR" for r in readings.values()):
            log.warning(
                "all %d channel(s) failed this tick; skipping emit", len(readings)
            )
            return

        self.emit(
            Event(
                type=EventType.CRYOSTAT_READING,
                driver=self.name,
                payload={"readings": readings},
            )
        )

    def _read_channel(self, channel: str, unit: str) -> dict[str, Any]:
        """Read a single value-tree channel from the Bluefors Control API.

        Mirrors the "values" endpoint example in the Bluefors reference: GET
        the endpoint path (channel with dots replaced by slashes) and read
        ``data.content.latest_valid_value``, falling back to
        ``latest_value`` if there is no recent valid sample.
        """
        url = f"{self.bluefors_base_url}/values/{channel.replace('.', '/')}"
        params = {"key": self.api_key} if self.api_key else {}

        try:
            resp = requests.get(url, params=params, timeout=self.timeout)
            resp.raise_for_status()
            content = resp.json()["data"]["content"]
            sample = (
                content.get("latest_valid_value") or content.get("latest_value") or {}
            )
            raw_value = sample.get("value")
            value = float(raw_value) if raw_value not in (None, "") else None
            status = sample.get("status", "UNKNOWN")
        except Exception:
            log.exception("failed to read channel %s", channel)
            return {"value": None, "unit": unit, "status": "ERROR"}

        return {"value": value, "unit": unit, "status": status}


def normalize_channels(channels: dict[str, str] | list[str] | None) -> dict[str, str]:
    """Turn a channel list/dict/None into the ``{path: unit}`` shape the driver uses."""
    if channels is None:
        return {}
    if isinstance(channels, dict):
        return dict(channels)
    return {channel: "" for channel in channels}


def parse_channels(raw: str) -> dict[str, str]:
    """Parse ``"path[:unit],path[:unit],..."`` into a channel->unit dict.

    Used by the CLI, which only has a single string option to work with.
    """
    channels: dict[str, str] = {}
    for part in raw.split(","):
        part = part.strip()
        if not part:
            continue
        path, _, unit = part.partition(":")
        channels[path.strip()] = unit.strip()
    return channels


def build_from_options(
    *,
    qpi_addr: str,
    token: str,
    name: str,
    ca_fingerprint: str,
    ca_file_path: str,
    recv_timeout_ms: int,
    options: dict[str, str],
) -> BlueforsGen1Driver:
    """Build a driver from the CLI's generic ``-o key=value`` options.

    Recognised keys: ``channels`` (required, ``path[:unit],...``), ``base_url``,
    ``api_key``, ``poll_interval``, ``timeout``. Raising ``ValueError`` lets the
    CLI report a bad option uniformly, without knowing anything Bluefors-specific.
    """
    channels = options.get("channels", "")
    if not channels:
        raise ValueError(
            "bluefors_gen1 needs a 'channels' option, e.g. "
            "-o channels=mapper.bf.tmc:K,mapper.bf.pmc:mbar"
        )
    return BlueforsGen1Driver(
        qpi_addr=qpi_addr,
        token=token,
        name=name,
        bluefors_base_url=options.get("base_url", "http://127.0.0.1:49099"),
        channels=parse_channels(channels),
        api_key=options.get("api_key", ""),
        poll_interval=float(options.get("poll_interval", DEFAULT_POLL_INTERVAL)),
        timeout=float(options.get("timeout", DEFAULT_TIMEOUT)),
        ca_fingerprint=ca_fingerprint,
        ca_file_path=ca_file_path,
        recv_timeout_ms=recv_timeout_ms,
    )


def run_monitor(
    *,
    device: str,
    options: dict[str, str],
    qpi_addr: str,
    token: str,
    name: str,
    ca_fingerprint: str,
    ca_file_path: str,
    recv_timeout_ms: int,
) -> None:
    """Run the Bluefors Gen. 1 monitor, config from -o options.

    The uniform runner the `monitor` operation registry dispatches to; *device*
    is always ``bluefors_gen1`` here. See :func:`build_from_options` for the keys.
    """
    build_from_options(
        qpi_addr=qpi_addr,
        token=token,
        name=name,
        ca_fingerprint=ca_fingerprint,
        ca_file_path=ca_file_path,
        recv_timeout_ms=recv_timeout_ms,
        options=options,
    ).run()
